Sources/AsyncHTTPClient/AsyncAwait/Transaction+StateMachine.swift

```diff
@@ -0,0 +1,603 @@
+                    // requestStreamState, producing is already paused/finished.
+                    self.state = .executing(context, requestStreamState, .waitingForStream(buffer, next: .error(error)))
+                    return .none
+
+                case .producing:
+                    // The request body stream is producing. We stop the request body stream, by
+                    // changing the request stream state to paused. This will lead to no further
+                    // produce calls. Once the next bytes are delivered.
+                    self.state = .executing(context, .paused, .waitingForStream(buffer, next: .error(error)))
+                    return .none
+                }
+
+            case .executing(let context, let requestStreamState, .buffering(let streamID, let buffer, next: .askExecutorForMore)),
+                 .executing(let context, let requestStreamState, .buffering(let streamID, let buffer, next: .endOfFile)):
+
+                switch requestStreamState {
+                case .initialized:
+                    preconditionFailure("Invalid state: \(self.state)")
+
+                case .paused, .finished:
+                    self.state = .executing(context, requestStreamState, .buffering(streamID, buffer, next: .error(error)))
+                    return .none
+
+                case .producing:
+                    self.state = .executing(context, .paused, .buffering(streamID, buffer, next: .error(error)))
+                    return .none
+                }
+
+            case .executing(let context, _, .waitingForRemote(let streamID, let continuation)):
+                // We are in response streaming. The response stream is waiting for the next bytes
+                // from the server. We can fail the call to `next` immediately.
+                self.state = .finished(error: error, streamID)
+                return .failResponseStream(continuation, error, context.executor)
+
+            case .finished(error: _, _):
+                // The request has already failed, succeeded, or the users is not interested in the
+                // response. There is no more way to reach the user code. Just drop the error.
+                return .none
+
+            case .executing(let context, _, .finished(let streamID, let continuation)):
+                self.state = .finished(error: error, streamID)
+                return .failResponseStream(continuation, error, context.executor)
+            }
+        }
+
+        // MARK: - Request -
+
+        enum StartExecutionAction {
+            case cancel(HTTPRequestExecutor)
+            case none
+        }
+
+        mutating func willExecuteRequest(_ executor: HTTPRequestExecutor) -> StartExecutionAction {
+            switch self.state {
+            case .initialized(let continuation), .queued(let continuation, _):
+                let context = ExecutionContext(
+                    executor: executor,
+                    allocator: .init(),
```

Comment:
Do we have any plan to properly construct an allocator?

Reply:
Yes, my plan is to expose the channel allocator on `HTTPRequestExecutor` in a follow up pr.